lib/listen/record.rb

```diff
@@ -46,16 +46,18 @@ def file_data(rel_path)
     end
 
     def dir_entries(rel_path)
-      subtree = if ['', '.'].include? rel_path.to_s
+      rel_path_s = rel_path.to_s
+      subtree = if ['', '.'].include?(rel_path_s)
         @tree
       else
-        @tree[rel_path.to_s] ||= _auto_hash
-        @tree[rel_path.to_s]
+        @tree[rel_path_s] ||= _auto_hash
       end
 
-      subtree.transform_values do |values|
-        # only get data for file entries
-        values.key?(:mtime) ? values : {}
+      subtree.each_with_object({}) do |(key, values), result|
+        # only return data for file entries
+        if values.respond_to?(:has_key?)
```

Comment:
What's the reason for this extra check?

Reply:
The root of this bug is that for folders the subtree has an array of hashes (that respond to `has_key?`). But for files, `subtree` is itself a hash. So if a folder gets renamed into a file, we were crashing here trying to treat a float (mtime) as if it were a hash.
So my hack here is to just treat the symptoms and skip over entries like that, if they happen. I'm sure there's a better answer, and it probably involves the comment higher up about deleting this entire class. But I wasn't able to take on a scope like that.

Reply:
Hmm, I think I could address the above more cleanly a few lines up. Maybe check if `subtree` is an Array at all? I _think_ in the file case it will itself be a Hash.

Reply:
Turns out my check for Array wouldn't work, because in fact the top-level object is a hash with the filename as key. The values in turn are hashes with symbols as keys (like `:mtime`) and scalar values.

Reply:
I personally think `respond_to?` is a hack in this case. Not sure if there is a better alternative without spending some time debugging the data. Maybe you can `pp` the top level out so we can see exactly what it looks like?

Reply:
Agreed--it's definitely a hack!

I certainly could replace `respond_to?` with `is_a?(Hash)`. Would that seem any less hacky to you? Over the years I've gone back and forth between asking objects the `is_a?` question vs the `respond_to?`. The latter is a bit more "duck-typey" but sometimes less clear.

Reply:
I'm guessing `transform_values` is more efficient too.

Reply:
Yes, `transform_values` is probably a bit more efficient. I'd be shocked if there were a case where it mattered here, though. The real cost of this gem is the multiple threads it creates and all the event reprocessing it does.